src/tasks/move_directory.rs

```diff
@@ -0,0 +1,177 @@
+use crate::core::{Core, Target};
+use crate::errors;
+use crate::tasks::Task;
+use std::{fs, path};
+use tracing_batteries::prelude::tracing;
+
+pub struct MoveDirectory {
+    pub new_path: path::PathBuf,
+}
+
+#[async_trait::async_trait]
+impl Task for MoveDirectory {
+    #[tracing::instrument(name = "task:move_directory(repo)", err, skip(self, _core))]
+    async fn apply_repo(
+        &self,
+        _core: &Core,
+        repo: &crate::core::Repo,
+    ) -> Result<(), crate::core::Error> {
+        if !repo.exists() {
+            return Err(errors::user(
+                format!("The repository {} does not exist on your machine and cannot be moved as a result.", repo.name).as_str(),
+                format!("Make sure the name is correct and that the repository exists by running `git-tool clone {}` first.", repo.name).as_str()
+            ));
+        }
+
+        fs::rename(repo.path.clone(), self.new_path.clone()).map_err(|err| {
+            errors::user_with_internal(
+                &format!(
+                    "Could not rename the repository directory '{}' to '{}' due to an OS-level error.",
+                    repo.path.display(),
+                    self.new_path.display()
+                ),
+                "Check that Git-Tool has permission to create this directory and any missing parent directories.",
+                err,
+            )
+        })?;
+
+        Ok(())
+    }
+
+    #[tracing::instrument(name = "task:git_rename(scratchpad)", err, skip(self, _core))]
+    async fn apply_scratchpad(
+        &self,
+        _core: &Core,
+        scratch: &crate::core::Scratchpad,
+    ) -> Result<(), crate::core::Error> {
+        if !scratch.exists() {
+            return Err(errors::user(
+                format!("The scratchpad {} does not exist on your machine and cannot be moved as a result.", scratch.get_name()).as_str(),
+                "Make sure the name is correct and that the scratchpad exists first."
+            ));
+        }
+
+        fs::rename(scratch.get_path(), self.new_path.clone()).map_err(|err| {
+            errors::user_with_internal(
+                &format!(
+                    "Could not rename the scratchpad directory '{}' to '{}' due to an OS-level error.",
+                    scratch.get_path().display(),
+                    self.new_path.display()
+                ),
+                "Check that Git-Tool has permission to create this directory and any missing parent directories.",
+                err,
+            )
+        })?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::tasks::GitClone;
+    use tempfile::tempdir;
+
+    #[tokio::test]
+    async fn move_directory_successfully() {
```

Comment:
This is gonna need `#[cfg_attr(feature = "pure-tests", ignore)]` too

Reply:
Refactored this test to be pure